vm/chef/cookbooks/singlefs/files/singlefs-install

```diff
@@ -0,0 +1,296 @@
+  # Fix issue related to ZFS exports without /etc/exports.
+  sed -i "s/if \[ -f \/etc\/exports \]/if \[ -f \/etc\/exports \] #/" \
+    /etc/init.d/nfs-kernel-server
+sed -i "s/ZFS_SHARE='no'/ZFS_SHARE='yes'/" /etc/default/zfs
+else
+  for network in "${networks[@]}"; do
+    echo "/${STORAGE_POOL} ${network}(rw,no_subtree_check,fsid=100)" >> /etc/exports
+  done
+fi
+
+if is_centos; then
+  NFS_CONFIG="/etc/sysconfig/nfs"
+  NFS_SERVICES=(nfs-server.service)
+  SAMBA_SERVICES=(smb.service nmb.service)
+fi
+if is_debian; then
+  NFS_CONFIG="/etc/default/nfs-kernel-server"
+  NFS_SERVICES=(nfs-kernel-server)
+  SAMBA_SERVICES=(smbd nmbd)
+  # Fix issue related to ZFS exports without /etc/exports.
+  sed -i "s/if \[ -f \/etc\/exports \]/if \[ -f \/etc\/exports \] #/" \
+    /etc/init.d/nfs-kernel-server
+fi
+
+# Use 16 * (# cores) NFS threads. This provides good performance for both
+# streaming and random IO without overwhelming the CPU with thread scheduling.
+readonly NFS_THREADS=$(($(nproc) * 16 ))
+sed -i "s/#*RPCNFSDCOUNT=.*/RPCNFSDCOUNT=${NFS_THREADS}/" "${NFS_CONFIG}"
+
+# Use a fixed port number for mountd, so that the port does not change
+# during failover.
+sed -i "s/RPCMOUNTDOPTS=\"/RPCMOUNTDOPTS=\"--port 2050 /" "${NFS_CONFIG}"
+
+# Start and enable NFS. Don't fail installation if NFS doesn't return
+# succesfully on start. The validation script checks if NFS started correctly.
+if [[ "${ENABLE_NFS}" = "enable:True" ]]; then
+  for service in "${NFS_SERVICES[@]}"; do
+    systemctl enable "${service}" || fail "Failed to enable service: ${service}"
+    systemctl start "${service}" || true
+  done
+fi
+
+# Configure SMB.
+cp /etc/samba/smb.conf /etc/samba/smb.conf.orig
+cat >> /etc/samba/smb.conf << EOF
+# Public share added by click to deploy for testing.
+# Remove this for production work loads.
+[${STORAGE_POOL}]
+   comment = Public share of /${STORAGE_POOL}
+   path = /${STORAGE_POOL}
+   browseable = yes
+   read only = no
+   writable = yes
+   guest ok = yes
+   hosts allow = ${networks[@]}
+   hosts deny = 0.0.0.0/0
+EOF
+
+# Enable SMB.
+if [[ "${ENABLE_SMB}" = "enable:True" ]]; then
+  for service in "${SAMBA_SERVICES[@]}"; do
+    systemctl enable "${service}" || fail "Failed to enable service: ${service}"
+    systemctl start "${service}" || fail "Failed to start service: ${service}"
+  done
+fi
+
+# Basic firewall configuration
+
+# Allow established sessions.
+iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
+
+# Allow ssh from the internet
+iptables -A INPUT -p tcp --dport ssh -j ACCEPT
+
+# Allow internal traffic on the local network, including NFS and SMB.
+for network in "${networks[@]}"; do
+  iptables -A INPUT -s "${network}" -j ACCEPT
+done
+
+# Block everything else
+iptables -P INPUT DROP
+
+# Save firewall rules
+if is_debian; then
+  iptables-save > /etc/iptables/rules.v4
+fi
+if is_centos; then
+  service iptables save
+fi
+
+echo "Software installation complete."
```

Comment:
Could we refactor this script by extracting some of its features to functions in external scripts.